addons/audio/alsa.c

```diff
@@ -875,6 +878,94 @@ static void alsa_deallocate_recorder(ALLEGRO_AUDIO_RECORDER *r)
    snd_pcm_close(data->capture_handle);
 }
 
+static void _device_list_dtor(void* value, void* userdata)
+{
+   (void)userdata;
+
+   ALLEGRO_AUDIO_DEVICE* device = (ALLEGRO_AUDIO_DEVICE*)value;
+   al_free(device->name);
+   al_free(device->identifier);
+}
+
+static _AL_LIST* alsa_get_devices(void)
+{
+   if (!device_list) {
+      device_list = _al_list_create();
+
+      int rcard = -1;
+      while(snd_card_next(&rcard) == 0) {
+         if (rcard < 0) return device_list;
+
+         snd_ctl_t *handle;
+         char str[64];
+
+         sprintf(str, "hw:%i", rcard);
+         if (snd_ctl_open(&handle, str, 0) < 0) {
+            return device_list;
```

Comment:
For these early returns etc, it's useful to add an ALLEGRO_DEBUG("snd_ctl_open failed %d with error code %d\n", rcard, error_code);